src/lints/enum_struct_variant_changed_kind.ron

```diff
@@ -0,0 +1,77 @@
+    reference_link: Some("https://doc.rust-lang.org/reference/items/enumerations.html"),
+    query: r#"
+    {
+        CrateDiff {
+            baseline {
+                item {
+                    ... on Enum {
+                        visibility_limit @filter(op: "=", value: ["$public"])
+                        enum_name: name @output @tag
+
+                        importable_path {
+                            path @output @tag
+                            public_api @filter(op: "=", value: ["$true"])
+                        }
+
+                        variant {
+                            ... on StructVariant {
+                                kind: __typename @output @tag
+                                public_api_eligible @filter(op: "=", value: ["$true"])
+                                attrs @filter(op: "not_contains", value: ["$non_exhaustive"])
+                                variant_name: name @output @tag
+
+                                field @fold @transform(op: "count") @filter(op: ">", value: ["$zero"]) {
+                                    public_api_eligible @filter(op: "=", value: ["$true"])
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            current {
+                item {
+                    ... on Enum {
+                        visibility_limit @filter(op: "=", value: ["$public"])
+                        name @filter(op: "=", value: ["%enum_name"])
+
+                        importable_path {
+                            path @filter(op: "=", value: ["%path"])
+                            public_api @filter(op: "=", value: ["$true"])
+                        }
+
+                        # Don't check for exhaustiveness and public API eligibility here.
+                        # Variants that changed kind and gained `#[doc(hidden)]` should
+                        # report for both lints, since there is no implied continuity
+                        # between the old variant and the new one with the same name.
+                        variant {
+                            name @filter(op: "=", value: ["%variant_name"])
+                            new_kind: __typename @filter(op: "!=", value: ["%kind"]) @output
+
+                            span_: span @optional {
+                                filename @output
+                                begin_line @output
+                                end_line @output
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }"#,
+    arguments: {
+        "public": "public",
+        "true": true,
+        "zero": 0,
+        "non_exhaustive": "#[non_exhaustive]",
+    },
+    error_message: "A pub enum's struct variant with at least one pub field has changed to a different kind of enum variant, breaking possible instantiations and field accesses.",
```

Comment:
We don't know if user code can create that variant (since we don't know if it's exhaustive or not), so let's not mention that. Accessing the pub field is a sufficient example of breakage.
```suggestion
    error_message: "A pub enum's struct variant with at least one pub field has changed to a different kind of enum variant, breaking access to its pub fields.",
```